src/react.ts

```diff
@@ -110,6 +104,11 @@ const createImpl = <T>(createState: StateCreator<T, [], []>) => {
   const api =
     typeof createState === 'function' ? createStore(createState) : createState
 
+  if (!api.getServerState) {
+    const initialState = api.getState()
+    api.getServerState = () => initialState
+  }
+
```

Comment:
this is what I had in mind with my comment here:

- https://github.com/pmndrs/zustand/pull/2277#issuecomment-1872232908

Reply:
I don't think this helps, as you said `initialState` is not really "initial". We shouldn't modify `api` here anyways.

Reply:
so my thinking was that without this, we would get hydration errors on a simple manual state restore from somewhere else, like:

```
const useStore = create((set) => ({
  bears: 0,
  increasePopulation: () => set((state) => ({ bears: state.bears + 1 })),
  removeAllBears: () => set({ bears: 0 }),
}))

// only on client
if (typeof window !== 'undefined') {
  const bears = window.localStorage.getItem('bears')
  useStore.setState(state => ({ ... state, bears }))
}
```

by "capturing" the initial state of `bears:0`, which was used to do the server-side rendering, we make sure that the first client side render will also render with `bears:0`.

The `persist` middleware has to do this too because otherwise, at this point, we don't know the initial state anymore. But doing it here also seems correct. Yes, when someone writes code like this, we could point them to the persist middleware, but the same thing could happen with a fast async restore (e.g. `localforage`) or really just calling `setState` on the client with something else (like a window global that was populated by the server or a third party script ...)

Reply:
Hmmmm, okay, if that's the case, I'd probably do it in vanilla and rename it to `getInitialState`. But, I'm not 100% sure if we go with it. Let me add a commit and see how it feels.